File: sdks/android/src/main/kotlin/com/likerts/sdk/Visibility.kt
package com.likerts.sdk

import kotlinx.serialization.json.*

private fun selected(answer:JsonElement):List<String>? = when(answer){is JsonArray->answer.mapNotNull{it.jsonPrimitive.contentOrNull};is JsonObject->answer["selected"]?.jsonArray?.mapNotNull{it.jsonPrimitive.contentOrNull};is JsonPrimitive->answer.contentOrNull?.let(::listOf)}
private fun answered(answer:JsonElement?):Boolean = when(answer){null,JsonNull->false;is JsonArray->answer.isNotEmpty();is JsonObject->selected(answer)?.isNotEmpty()==true;is JsonPrimitive->if(answer.isString)answer.content.trim().isNotEmpty()else true}
internal fun conditionMatches(condition:VisibilityCondition,answer:JsonElement?):Boolean = when(condition.operator){
    "answered"->answered(answer);"not_answered"->!answered(answer)
    "equals","not_equals"->{if(answer==null)false else {val actual=selected(answer)?.takeIf{it.size==1}?.first()?.let(::JsonPrimitive)?:answer;val equal=actual==condition.value;if(condition.operator=="equals")equal else !equal}}
    "includes","not_includes"->{if(answer==null)false else {val included=condition.value?.contentOrNull?.let{selected(answer)?.contains(it)}==true;if(condition.operator=="includes")included else !included}}
    else->throw IllegalArgumentException("Unsupported visibility operator")
}
fun visibleQuestionIds(questions:List<Question>,answers:Map<String,JsonElement>):Set<String>{val byId=questions.associateBy{it.id};val memo=mutableMapOf<String,Boolean>();val active=mutableSetOf<String>();fun visible(q:Question):Boolean{memo[q.id]?.let{return it};check(active.add(q.id)){"Conditional visibility cycle"};val condition=q.visibleWhen;val result=if(condition==null)true else {val source=requireNotNull(byId[condition.questionId]){"Conditional visibility references an unknown question"};conditionMatches(condition,if(visible(source))answers[source.id]else null)};active.remove(q.id);memo[q.id]=result;return result};return questions.filter(::visible).mapTo(mutableSetOf()){it.id}}
fun visibleAnswers(questions:List<Question>,answers:Map<String,JsonElement>):Map<String,JsonElement>{val visible=visibleQuestionIds(questions,answers);return answers.filterKeys(visible::contains)}

File: sdks/ios/Sources/Likerts/Visibility.swift
import Foundation

extension Answer {
  var selectedValues:[String]? { switch self { case .text(let value): return [value];case .choices(let values),.ranking(let values): return values;case .other(let selected,_):return selected;case .number,.matrixSingle,.matrixMultiple,.allocation:return nil } }
  var isAnswered:Bool { switch self {case .text(let value):return !value.trimmingCharacters(in:.whitespacesAndNewlines).isEmpty;case .number:return true;case .choices(let values),.ranking(let values):return !values.isEmpty;case .other(let selected,_):return !selected.isEmpty;case .matrixSingle(let values):return !values.isEmpty;case .matrixMultiple(let values):return !values.isEmpty;case .allocation(let values):return !values.isEmpty} }
}

func conditionMatches(_ condition:VisibilityCondition,_ answer:Answer?)->Bool {
  switch condition.operator {
  case .answered:return answer?.isAnswered==true
  case .notAnswered:return answer?.isAnswered != true
  case .equals,.notEquals:
    guard let answer else{return false};let equal:Bool
    switch (answer,condition.value) {case (.number(let actual),.number(let expected)):equal=actual==expected;case (_, .string(let expected)):equal=answer.selectedValues==[expected];default:equal=false}
    return condition.operator == .equals ? equal : !equal
  case .includes,.notIncludes:
    guard let answer,case .string(let expected)?=condition.value else{return false};let included=answer.selectedValues?.contains(expected)==true
    return condition.operator == .includes ? included : !included
  }
}

public func visibleQuestionIDs(_ questions:[Question],answers:[String:Answer])->Set<String> {
  let byId=Dictionary(uniqueKeysWithValues:questions.map{($0.id,$0)});var memo:[String:Bool]=[:];var active=Set<String>()
  func visible(_ question:Question)->Bool {if let value=memo[question.id]{return value};if !active.insert(question.id).inserted{return false};defer{active.remove(question.id)};guard let condition=question.visibleWhen else{memo[question.id]=true;return true};guard let source=byId[condition.questionId] else{memo[question.id]=false;return false};let result=conditionMatches(condition,visible(source) ? answers[source.id]:nil);memo[question.id]=result;return result}
  return Set(questions.filter(visible).map(\.id))
}

public func visibleAnswers(_ questions:[Question],answers:[String:Answer])->[String:Answer] {
  let visible=visibleQuestionIDs(questions,answers:answers);return answers.filter{visible.contains($0.key)}
}
